examples/graphs/node2vec_movielens.py

```diff
@@ -0,0 +1,563 @@
+2. **In-out parameter** (`q`): Allows the search to differentiate
+between *inward* and *outward* nodes. Setting it to a high value biases the
+random walk towards local nodes, while setting it to a low value biases the walk
+to visit nodes which are further away.
+
+"""
+
+
+def next_step(graph, previous, current, p, q):
+    neighbors = list(graph.neighbors(current))
+
+    weights = []
+    # Adjust the weights of the edges to the neighbors with respect to p and q.
+    for neighbor in neighbors:
+        if neighbor == previous:
+            # Control the probability to return to the previous node.
+            weights.append(graph[current][neighbor]["weight"] / p)
+        elif graph.has_edge(neighbor, previous):
+            # The probability of visiting a local node.
+            weights.append(graph[current][neighbor]["weight"])
+        else:
+            # Control the probability to move forward.
+            weights.append(graph[current][neighbor]["weight"] / q)
+
+    # Compute the probabilities of visiting each neighbor.
+    weight_sum = sum(weights)
+    probabilities = [weight / weight_sum for weight in weights]
+    # Probabilistically select a neighbor to visit.
+    next = np.random.choice(neighbors, size=1, p=probabilities)[0]
+    return next
+
+
+def randm_walk(graph, num_walks, num_steps, p, q):
+    walks = []
+    nodes = list(graph.nodes())
+    # Perform multiple iterations of the random walk.
+    for walk_iteration in range(num_walks):
+        random.shuffle(nodes)
+
+        for node in tqdm(
+            nodes,
+            position=0,
+            leave=True,
+            desc=f"Random walks iteration {walk_iteration + 1} of {num_walks}",
+        ):
+            # Start the walk with a random node from the graph.
+            walk = [node]
+            # Randomly walk for num_steps.
+            while len(walk) < num_steps:
+                current = walk[-1]
+                previous = walk[-2] if len(walk) > 1 else None
+                # Compute the next node to visit.
+                next = next_step(graph, previous, current, p, q)
+                walk.append(next)
+            # Replace node ids (movie ids) in the walk with token ids.
+            walk = [vocabulary_lookup[token] for token in walk]
+            # Add the walk to the generated sequence.
+            walks.append(walk)
+
+    return walks
+
+
+"""
+## Generate Training Data using Random Walk
+You can explore different configurations of `p` and `q` to different results of
+related movies.
+"""
+
+p = 1
+q = 1
+num_walks = 5
+num_steps = 10
+walks = randm_walk(movies_graph, num_walks, num_steps, p, q)
+
+print("Number of walks generated:", len(walks))
+
+"""
+## Generate Positive and Negative Examples
+
+To train a skip-gram model, we use the generated walks to create positive and
+negative training examples. Each example includes the following features:
+1. `target`: A movie in a walk sequence.
```

Comment:
Introduce a line break above any list so that it renders properly

Reply:
done